tests/test_utils/test_format_file_name.py

```diff
@@ -0,0 +1,59 @@
+import datetime
+import pytest
+
+def format_filename(format: str, data: dict) -> str:
+    """
+    Formats the filename based on the provided template and variables.
+    """
+    if not isinstance(format, str):
+        raise TypeError("format argument must be a string")
+
+    filename = format.format(**data)
+    filename += data["ext"]
+    return filename
+
+base_data = {
+    "original": "file",
+    "number": 1,
+    "algorithm": "alg",
+    "date": datetime.date(
+        2023,
+        1,
+        23,
+    ),
+    "time": datetime.time(12, 30, 45),
+    "ext": ".jpg",
+}
+
+
+# Tests
+@pytest.mark.parametrize(
+    "format_str, data, expected",
+    [
+        ("{original}_{algorithm}", base_data, "file_alg.jpg"),
+        ("{original}_{number}_{date}", base_data, "file_1_2023-01-23.jpg"),
+        ("", base_data, ".jpg"),
+        ("{original}", {}, KeyError),
+        ("{nonexistent_key}", base_data, KeyError),
+        (123, base_data, TypeError),
+        ("{original}_@#$_{algorithm}", base_data, "file_@#$_alg.jpg"),
+        ("{original}" * 100, base_data, "file" * 100 + ".jpg"),
+        (
+            "{original}_{number}",
+            {"original": "file", "number": 123, "ext": ".jpg"},
+            "file_123.jpg",
+        ),
+        (
+            "{date}_{time}",
+            {"date": "2023/01/23", "time": "12-00", "ext": ".log"},
+            "2023/01/23_12-00.log",
+        ),
+        ("{original}", {"original": "file", "ext": ""}, "file"),
+    ],
+)
+def test_format_filename(format_str, data, expected):
+    if isinstance(expected, type) and issubclass(expected, Exception):
```

Comment:
Its bad practice to do generic exception catching.  The command line will display ValueErrors as messages to the user about what mistake they made.  So catch expected errors and the re-raise more descriptive useful ValueError.

Reply:
Will change the exception here.

Reply:
Separated exception handling to distinguish between expected and unexpected Exceptions.